update general targetRef policy docs (#1131)
kumahq/kuma-website: app/docs/dev/policies/targetref.md

---
title: Understanding TargetRef policies
---

## What is a policy?

> A policy is anything that can modify the configuration of a proxy.

Reviewer: This definition is a little vague for me. If I were unfamiliar with Kuma and it's policies, I wouldn't understand this definition. [Here](https://kuma.io/docs/1.8.x/explore/dpp/#dynamic-configuration-of-the-data-plane-proxy) docs say that policies are used to generate DP configuration. So maybe something like this: 
`A policy is a set of guidelines that will be used to build proxy configuration. Kuma combines policies and Dataplane entities to build configuration for proxy`

## What do `targetRef` policies look like?

> There are 2 useful parts in a policy:

Reviewer: `useful parts` suggests that these are not mandatory fields. Are they?

In particular, it defines which proxies have their Envoy configuration modified.

Some policies also support further narrowing.

The `spec.to[].targetRef` field defines rules that applies to outgoing traffic of proxies selected by `spec.targetRef`.
The `spec.from[].targetRef` field defines rules that applies to incoming traffic of proxies selected by `spec.targetRef`.

The actual configuration is defined in a `default` map.

For example:

```yaml
type: ExamplePolicy
name: my-example
mesh: default
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
        kind: Mesh
      default: # Configuration that applies to outgoing traffic
        key: value
  from:
    - targetRef:
        kind: Mesh
      default: # Configuration that applies to incoming traffic
        key: value
```

Some policies are not directional and will not have `to` and `from`.
For example

```yaml
type: NonDirectionalPolicy
name: my-example
mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    key: value
```

{% tip %}
One of the benefits of `targetRef` policies is that the spec is always the same between Kubernetes and Universal.

> This means that converting policies between universal and Kubernetes only means rewriting the metadata.

Reviewer 2: ```suggestion
This means that converting policies between Universal and Kubernetes only means rewriting the metadata.
```

{% endtip %}

#### Writing a `targetRef`

`targetRef` is a concept borrowed from [Kubernetes Gateway API](https://gateway-api.sigs.k8s.io/) its usage is fully defined in [MADR 005](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md).
Its goal is to select subsets of proxies with maximum flexibility.

It looks like:

```yaml
targetRef:
  kind: Mesh | MeshSubset | MeshService | MeshServiceSubset | MeshGatewayRoute

>   name: "my-name" # For kinds MeshService, MeshServiceSubset and MeshGatewayRoute a name can be defined

Reviewer 2: Why can't all of them have a name?

Author: Cause for Mesh/MeshSubset you don't have a choice right now (it must be the one that the policy is in) In the future we might relax this but not now

      default:
        backends:
          - file:
              format:
                plain: '{"start_time": "%START_TIME%"}'
              path: "/tmp/logs.txt"
```

Using `spec.targetRef`, this policy targets all proxies that implement the service `web-frontend`.
It defines the scope of this policy as applying to traffic either from or to `web-frontend` services.

The `spec.to.targetRef` section enables logging for any traffic going to `web-backend`.
The `spec.from.targetRef` section enables logging for any traffic coming from _any service_ in the `Mesh`.

### Target resources

Not every policy supports `to` and `from` levels. Additionally, not every resource can
appear at every supported level. The specified top level resource can also affect which
resources can appear in `to` or `from`.

To help users, each policy documentation includes a table indicating which `targetRef` kinds is supported at each level.

This table looks like:

| `targetRef.kind`    | top level | to  | from |
| ------------------- | --------- | --- | ---- |
| `Mesh`              | ✅        | ✅  | ❌   |
| `MeshSubset`        | ✅        | ❌  | ❌   |
| `MeshService`       | ✅        | ❌  | ✅   |

> | `MeshServiceSubset` | ✅        | ❌  | ❌   |

Reviewer 2: We mention MeshGatewayRoute on line [152](https://github.com/kumahq/kuma-website/pull/1131/files#diff-f25d24e526ca4764872874b844778d34817a88616a050f4962aa713fc50e5c08R152) but not here, should we make this consistent?

Here it indicates that the top level can use any targetRef kinds. But in `targetRef.to` only kind `Mesh` can be used and in `targetRef.from` only kind `MeshService`.

### Merging configuration

> Because for a specific proxy multiple `targetRef` can be used.
> It is necessary to define a policy for merging configuration.

Reviewer 2: ```suggestion
It is necessary to define a strategy for merging configurations, 
because a proxy can be targeted by multiple `targetRef`s.
```

> We define a total order this way:

Reviewer 2: ```suggestion
We define a total order:
```

> - Mesh > MeshSubset > MeshService > MeshServiceSubset > MeshGatewayRoute (The more a `targetRef` is specific the higher priority it has)

Reviewer 2: I would not use "specific" here because it was used in the S&D explanation.

```suggestion
- Mesh > MeshSubset > MeshService > MeshServiceSubset > MeshGatewayRoute (the more a `targetRef` is focused the higher priority it has)
```

- If levels are equal the lexicographic order of policy names is used

For `to` and `from` policies we concatenate the array for each matching policies.
We then build configuration by merging each level using [JSON patch merge](https://www.rfc-editor.org/rfc/rfc7386).

For example if I have 2 `default` ordered this way:

```yaml
conf: 1
sub:
  array: [1, 2, 3]
  other: 50
  other-array: [3, 4, 5]
---
sub:
  array: []
  other: null
  other-array: [5, 6]
  extra: 2
```

The merge result is:
```yaml
conf: 1
sub:
  array: []
  other-array: [5, 6]
  extra: 2

> ```

Reviewer 2: Can we have this prefixed with `spec` or some other way link this to the actual policy "world" so that the user is familiar with the constructs?

### Examples

> #### Applying as a global default

Reviewer 2: ```suggestion
#### Applying a global default
```
